### DISCUSSION TITLE
Include partial but only a specific line range

### DISCUSSION THREAD

**Discussion by ntgussoni** [markdoc/markdoc#133] [Ideas]
Hello guys, 

We are evaluating moving away from asciidoc, but I want to ensure we have support for all the current features we are using.

One of the features is including local or external files (kind of like partials) but where you can also specify a line range to include.
https://docs.asciidoctor.org/asciidoc/latest/directives/include-lines/

Would it make sense to ship this as part of the core or do you see it being something external (AKA implement it yourself).

I can imagine something like this

```handlebars
{% partial file="http::/www.foo.bar/my/file.anything lines=[1-3, 4-8, 10] %}
```

- The partial would try to download it if it's an external FQDN (I guess we need an `unsafe: true` attribute passed to .parse() for safety reasons, so that by default we never allow including external files that are not in the file system).
- After downloading the file, we'd pick up the line ranges specified.

I understand there are some things to consider (and I'm probably missing some cases) but correct me if I'm wrong.
- This has to be done prior to the AST conversion, so at the beginning of `.parse`
- What happens with a file that contains frontmatter?

Thanks for your input and work! I'm very excited about markdoc

**Comment by ntgussoni** [comment]
It could also be even trickier, and I think this is not how the partial works, we would need to replace the `{% partial ...%}` tag with the contents in the range, and then run `.parse()` on the host file.

current wow with partials involves generating the AST, adding it to `config.partials` so that it gets embedded once you parse your "host" file.

**Reply by mfix-stripe** [reply]
You would have to create your own custom tag for this (which means I would probably call it something else besides partial, just so you can use built in partials as well), but that should be possible. I would recommend scanning through your ASTs at build time, pulling out the requested URLs, and caching the content in a manifest of some sort—this way your runtime is ~instant.

**Reply by ntgussoni** [reply]
Yeah, that makes sense.  I'll give it some thought time 

Just to be clear whether to ask these questions or not, are you expecting to have a full range of components as part of this project or do you expect people to just extend markdoc?

thanks for the time!

**Reply by mfix-stripe** [reply]
The vast majority of the time we expect folks to extend Markdoc with their own custom tags, nodes, functions, renderers, etc. On rare occasions, we might add new tags to the core, but this will be to enable functionality that we want to remain consistent, like the `if` and `else` tag.
